### ISSUE TITLE
Speed up normalise part of process_topdecs

### ISSUE THREAD

**Issue by xrchz** [CakeML/cakeml#229]
Currently, `process_topdecs` and `parse_topdecs` in `cfTactics(Base)Lib` are implemented in terms of `EVAL`. However, the theorem produced is not required except for its right-hand-side. This suggests it would be better to run the parser outside the logic. Running it with `EVAL` is a noticeable speed bottleneck.

**Comment by mn200** [comment]
A little investigation has revealed that the main slowdown is called by the CF framework's normalise function.

**Comment by xrchz** [comment]
@Armael did you get anywhere with speeding up the normaliser?

**Comment by Armael** [comment]
I did some progress, but I need to wrap everything up and finish the thing. I'll try to get that done next week...

**Comment by xrchz** [comment]
Just noting that the current implementation of `normalise` can pick bad names that clash with the user's names for `let` variables. So when you implement the faster version, it could be good to fix that too...

**Comment by xrchz** [comment]
I think it's because `free_in` is not fully implemented.

**Comment by Armael** [comment]
Do you have an example of such case where not-fresh-enough names are used?

**Comment by xrchz** [comment]
Here's a simple one
```
val bug = process_topdecs`
  fun foo a = let val b = a in ((a,1),b) end`;
```

**Comment by Armael** [comment]
Indeed. I've just pushed a fix to both the HOL and ML normalisers (in the `mlnormalise` branch) (the fix to the HOL normaliser can be cherry-picked to master if needed).

**Comment by Armael** [comment]
This can be closed once #179 is merged (I think).

**Comment by xrchz** [comment]
merged
